internal/transport/grpcframer/framer.go

```diff
@@ -0,0 +1,267 @@
+ * limitations under the License.
+ *
+ */
+
+// Package grpcframer defines the interface for implementing an HTTP/2 framer,
+// its required types and an implementation of said framer.
+package grpcframer
+
+import "sync"
+
+// FrameType represents the type of a Frame and its value according to the
+// HTTP/2 spec.
+// See https://httpwg.org/specs/rfc7540.html#FrameTypes.
+type FrameType uint8
+
+const (
+	DataFrameType         FrameType = 0x0
+	HeadersFrameType      FrameType = 0x1
+	PriorityFrameType     FrameType = 0x2
+	RSTStreamFrameType    FrameType = 0x3
+	SettingsFrameType     FrameType = 0x4
+	PushPromiseFrameType  FrameType = 0x5
+	PingFrameType         FrameType = 0x6
+	GoAwayFrameType       FrameType = 0x7
+	WindowUpdateFrameType FrameType = 0x8
+	ContinuationFrameType FrameType = 0x9
+)
+
+// Flags represents the flags that can be set on every frame type.
+type Flags uint8
+
+const (
+	// Data Frame
+	FlagDataEndStream Flags = 0x1
+	FlagDataPadded    Flags = 0x8
+
+	// Headers Frame
+	FlagHeadersEndStream  Flags = 0x1
+	FlagHeadersEndHeaders Flags = 0x4
+	FlagHeadersPadded     Flags = 0x8
+	FlagHeadersPriority   Flags = 0x20
+
+	// Settings Frame
+	FlagSettingsAck Flags = 0x1
+
+	// Ping Frame
+	FlagPingAck Flags = 0x1
+
+	// Continuation Frame
+	FlagContinuationEndHeaders Flags = 0x4
+
+	FlagPushPromiseEndHeaders Flags = 0x4
+	FlagPushPromisePadded     Flags = 0x8
+)
+
+// Setting represents the id and value pair of a particular HTTP/2 setting.
+// See https://httpwg.org/specs/rfc7540.html#SettingFormat.
+type Setting struct {
+	ID    SettingID
+	Value uint32
+}
+
+// SettingID represents the id of a specific HTTP/2 setting.
+// See https://httpwg.org/specs/rfc7540.html#SettingValues.
+type SettingID uint16
+
+const (
+	SettingsHeaderTableSize      SettingID = 0x1
+	SettingsEnablePush           SettingID = 0x2
+	SettingsMaxConcurrentStreams SettingID = 0x3
+	SettingsInitialWindowSize    SettingID = 0x4
+	SettingsMaxFrameSize         SettingID = 0x5
+	SettingsMaxHeaderListSize    SettingID = 0x6
+)
+
+// FrameHeader is the 9 byte header of any HTTP/2 Frame.
+// See https://httpwg.org/specs/rfc7540.html#FrameHeader.
+type FrameHeader struct {
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+
+	// Type is a byte that represents the Frame Type. The HTTP/2 spec
+	// defines 10 standard types but extension frames may be written.
+	Type FrameType
+
+	// Flags is a byte representing the flags set for the specific frame
+	// type.
+	Flags Flags
+
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents any kind of HTTP/2 Frame, which can be casted into
+// its corresponding type using the Header method provided.
```

Comment:
Drop the part after comma. "Frame represents a HTTP/2 Frame" is more terse and makes more sense here. In general, comments are meant to explain the type, not it's different use cases and how it will be used throughout the codebase.